### PR TITLE
GH-174 - analyzing usages of WithAnyArgs/ForAnyArgs like methods

### LINKED ISSUE #174: Detect ReceivedWithAnyArgs/DidNotReceiveWithAnyArgs misuse
Could you detect invalid usages of `ReceivedWithAnyArgs`?

E.g.:

### Wrong `ReceivedWithAnyArgs` usage with `Arg.Is` -> this will not be checked 
```csharp
public interface ITest
{
    void DoSomething(string? thing);
}

public class Test
{
    [Test]
    public void TestSth()
    {

        var test = Substitute.For<ITest>();
        test.DoSomething("else");
        test.ReceivedWithAnyArgs().DoSomething(Arg.Is<string>(x => x.Contains("wrong")));
        test.Received().DoSomething(Arg.Is<string>(x => x.Contains("wrong"))); // correct -> fails test
    }
}
```

### Wrong `DidNotReceiveWithAnyArgs` usage with `Arg.Is` -> this will not be checked 
```csharp
        var test = Substitute.For<ITest>();
        test.DoSomething("this");
        test.DidNotReceive().DoSomething(Arg.Is<string>(x => x.Contains("that"))); // correct, does not fail test
        test.DidNotReceiveWithAnyArgs().DoSomething(Arg.Is<string>(x => x.Contains("that"))); // wrong, always fails
```

### PR DESCRIPTION
**Pull request by tpodolak**
closes #174

### CODE REVIEW

File: src/NSubstitute.Analyzers.Shared/Resources.resx [hunk 1]
```diff
@@ -375,6 +375,19 @@
     <value>Argument matcher used with a non-virtual member of a class.</value>
         <comment>The title of the diagnostic.</comment>
   </data>
+  
+      <data name="WithAnyArgsArgumentMatcherUsageDescription" xml:space="preserve">
+    <value>Argument matcher used with WithAnyArgs method.</value>
```
Comment (tpodolak) [inline src/NSubstitute.Analyzers.Shared/Resources.resx:380]: @dtchepak do you have any better alternative for warning title/description. IMO this one is not good enough

File: documentation/rules/NS5004.md [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+</tr>
+<tr>
+  <td>Category</td>
+  <td>Usage</td>
+</tr>
+</table>
+
+## Cause
+
+Argument matcher used with WithAnyArgs method.
```
Comment (tpodolak) [inline documentation/rules/NS5004.md:16]: See https://github.com/nsubstitute/NSubstitute.Analyzers/pull/189/files#r945298742

File: src/NSubstitute.Analyzers.VisualBasic/DiagnosticAnalyzers/SubstitutionNodeFinder.cs [hunk 3]
```diff
@@ -17,6 +18,7 @@ private SubstitutionNodeFinder()
     {
     }
 
+    // TODO replace with SyntaxVisitor or OperationVisitor
```
Comment (tpodolak) [inline src/NSubstitute.Analyzers.VisualBasic/DiagnosticAnalyzers/SubstitutionNodeFinder.cs:21]: This will be handled in separate PR. Most of the work to move to operation API will be done as part of https://github.com/nsubstitute/NSubstitute.Analyzers/issues/153

File: src/NSubstitute.Analyzers.Shared/Resources.resx (hunk 1 above)
Comment (dtchepak) [reply src/NSubstitute.Analyzers.Shared/Resources.resx:380]: Only thing I can think of is providing more information on the impact of this. Maybe something like:

```
Argument matcher used with WithAnyArgs. This matcher may not be executed.
```

?

File: documentation/rules/NS5004.md [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+To fix a violation of this rule, replace arg matchers used in aforementioned methods with `Arg.Any`
+
+For example:
+
+````c#
+// Incorrect:
+sub.DidNotReceiveWithAnyArgs().Bar(Arg.Is(1));
+
+// Correct:
+sub.DidNotReceiveWithAnyArgs().Bar(Arg.Any<int>());
```
Comment (dtchepak) [inline documentation/rules/NS5004.md:33]: Maybe add another correct example:

```
// Correct:
sub.DidNotReceiveWithAnyArgs().Bar(Arg.Any<int>());

// The following are also correct. The exact arguments will not be checked due to *WithAnyArgs.
sub.DidNotReceiveWithAnyArgs().Bar(default(int));
sub.DidNotReceiveWithAnyArgs().Bar(123);
```

Comment (tpodolak) [reply documentation/rules/NS5004.md:33]: Added

File: src/NSubstitute.Analyzers.Shared/Resources.resx (hunk 1 above)
Comment (tpodolak) [reply src/NSubstitute.Analyzers.Shared/Resources.resx]: Looks better. Changed. 
One more thing though. `WithAnyArgs` naming is not 100% accurate, as same rules apply to `ReturnsForAnyArgs`, `WhenForAnyArgs` etc. Should we come up with some more generic naming or we are fine with `WithAnyArgs`?

**Review by dtchepak — APPROVED** [review]
